packages/apollo-server-core/src/utils/schemaManager.ts

```diff
@@ -0,0 +1,216 @@
+        unsubscribeFromGateway?: Unsubscriber;
+      }
+    | {
+        readonly mode: 'schema';
+        readonly apiSchema: GraphQLSchema;
+      };
+
+  constructor(
+    options: (
+      | { gateway: GatewayInterface; apolloConfig: ApolloConfig }
+      | { apiSchema: GraphQLSchema }
+    ) & {
+      logger: Logger;
+      schemaDerivedDataProvider: SchemaDerivedDataProvider;
+    },
+  ) {
+    this.logger = options.logger;
+    this.schemaDerivedDataProvider = options.schemaDerivedDataProvider;
+    if ('gateway' in options) {
+      this.modeSpecificState = {
+        mode: 'gateway',
+        gateway: options.gateway,
+        apolloConfig: options.apolloConfig,
+      };
+    } else {
+      this.modeSpecificState = {
+        mode: 'schema',
+        apiSchema: options.apiSchema,
+      };
+      // The caller of the constructor expects us to fail early if the schema
+      // given is invalid/has errors, so we call the provider here.
+      options.schemaDerivedDataProvider(options.apiSchema);
+    }
+  }
+
+  /**
+   * Calling start() will:
+   * - Start gateway schema fetching (if a gateway was provided).
+   * - Initialize schema-derived data.
+   * - Synchronously notify onSchemaLoadOrUpdate() listeners of schema load, and
+   *   asynchronously notify them of schema updates.
+   */
+  public async start(): Promise<void> {
+    if (this.modeSpecificState.mode === 'gateway') {
+      const gateway = this.modeSpecificState.gateway;
+      if (gateway.onSchemaLoadOrUpdate) {
+        // Use onSchemaLoadOrUpdate if available, as it reports the core
+        // supergraph SDL and always reports the initial schema load.
+        this.modeSpecificState.unsubscribeFromGateway =
+          gateway.onSchemaLoadOrUpdate((schemaContext) => {
+            this.processSchemaLoadOrUpdateEvent(schemaContext);
+          });
+      } else {
+        this.modeSpecificState.unsubscribeFromGateway = gateway.onSchemaChange(
+          (apiSchema) => {
+            this.processSchemaLoadOrUpdateEvent({ apiSchema });
+          },
+        );
+      }
+
+      const config = await this.modeSpecificState.gateway.load({
+        apollo: this.modeSpecificState.apolloConfig,
+      });
+
+      // Note that for old gateways that have onSchemaChange() and no
+      // onSchemaLoadOrUpdate(), this.schemaDerivedData may not be initialized
+      // during gateway.load() (because old gateways don't notify listeners on
+      // schema load in some cases), so we must initialize it here if needed.
+      if (!this.schemaDerivedData) {
+        this.processSchemaLoadOrUpdateEvent({ apiSchema: config.schema });
+      }
+    } else {
+      this.processSchemaLoadOrUpdateEvent({
+        apiSchema: this.modeSpecificState.apiSchema,
+      });
+    }
+  }
+
+  /**
+   * Registers a listener for schema load/update events. Note that the latest
+   * event is buffered, i.e.
+   * - If registered before start(), then the callback will be first called in
```

Comment:
This isn't actually possible in the way ApolloServer uses it, right? We could just make it throw?

Reply:
Yep, `this.state` is private so users can't publicly call `onSchemaLoadOrUpdate()` before `start()`, and we don't call `onSchemaLoadOrUpdate()` before `start()` in our own code. I've changed the code in `onSchemaLoadOrUpdate()` to throw when called before `start()`.